### DISCUSSION TITLE
How i can be sure that no one else can benefit from the azure function which integrate with Chat GPT using API key

### DISCUSSION THREAD

**Discussion by mvcsharepointdev** [pnp/sp-dev-fx-webparts#3582] [Q&A]
I want to install the SPfx for chat GPT from this link @ https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-chatgpt-app

The steps requires creating and Azure Function which will store the API key are:-

OpenAI-Azure-functions

**Deploy**

To deploy you can use the Azure Extensions for VSCODE , go to "WORKSPACE" and on top Options select Deploy.

![image](https://user-images.githubusercontent.com/79726511/224542231-f1d61917-5a94-4d9d-88e2-d94681582481.png)

**Secure Azure Function App**

The Azure Fucntion is secured by Authentication after deployed go to Azure and select Function App and select the Azure Function App created "OpenAIFunctionsApp"

![image](https://user-images.githubusercontent.com/79726511/224542246-9df9e733-c956-42e3-8de2-45fac769cbda.png)

**configure authentication**

Select Authentications and you have option to create a new Azure App or select a one already created.

in this sample , create a new one.

![image](https://user-images.githubusercontent.com/79726511/224542250-7133780e-04a6-496d-beda-40da8ba4c7db.png)

After go to CORS and add your SharePoint address.

![image](https://user-images.githubusercontent.com/79726511/224542255-c42bafca-2b55-4ce4-9ecc-ffe6f8f621fb.png)

**Application Settings**

The function needs to have the followed environment vars defined:

![image](https://user-images.githubusercontent.com/79726511/224542256-ef36b320-f4dd-4cb6-9fd2-be6938608b28.png)

So at the end we will have an azure function which stores our secure API key. so how we can be sure that no external user can use/benefit from our azure function to create an own SPFx which call out azure function ?

Thanks

**Comment by vikramhn** [comment]
Could this answer your question?
https://learn.microsoft.com/en-us/sharepoint/dev/spfx/use-aadhttpclient-enterpriseapi

**Comment by bcameron1231** [comment]
As you described, the Azure Function is secured with Azure AD. That means only users in your organization (Inside your Azure AD) can use this. 

If you want to limit other SPFx Web Parts from being able to  able to use this function, you should convert the solution to be an Isolated Web Part, and only grant permission for this Web  Part to have access to this Azure Function App Registration.

https://learn.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/isolated-web-parts

Additionally, it may be smart to move your secret key from the Azure Function, to a Keyvault which has limited access. That way you don't accidentally expose this key to people in your Org with Azure access.

https://learn.microsoft.com/en-us/azure/app-service/app-service-key-vault-references?tabs=azure-cli
